vllm/v1/guided_decoding/__init__.py

```diff
@@ -0,0 +1,208 @@
+            self.requests = {
+                req
+                for req in self.requests if req.request_id not in request_ids
+            }
+
+    def populate_cache(self, request: Request):
+        if not request.use_guided_decoding:
+            return False
+        grammar = self.request_key_to_grammar.get(request.guided_decoding_key)
+        if grammar:
+            request.grammar = copy.copy(grammar)
+            return False
+        request.grammar = self.cache(request)
+        return True
+
+    def cache(self, request: Request):
+        return self.executor.submit(self._executor_loop, request)
+
+    def _executor_loop(self, request: Request) -> Grammar:
+        key = request.guided_decoding_key
+        with self._requests_lock:
+            self.requests.add(request)
+        if key in self.request_key_to_grammar:
+            grammar = self.request_key_to_grammar[key]
+            return copy.copy(grammar)
+        grammar = self.initialize_grammar(key)
+        # If cache is full, remove the least recently used item
+        if len(self.request_key_to_grammar) >= self.max_cache_size:
+            self.request_key_to_grammar.popitem(last=False)
+        self.request_key_to_grammar[key] = grammar
+        return copy.copy(grammar)
+
+    def initialize_grammar(self, key: GuidedDecodingKey) -> Grammar:
+        # Note that the request was validated in the engine core client,
+        # so at this point we know it is a supported type of request.
+        #
+        # TODO: we still need to handle xgrammar compilation failures
+        request_type, grammar_spec = key
+
+        if request_type == GuidedDecodingOptions.json:
+            # TODO -- allow any_whitespace to be configurable
+            # pending merge of https://github.com/vllm-project/vllm/pull/12744
+            ctx = self.compiler.compile_json_schema(grammar_spec,
+                                                    any_whitespace=False)
+        elif request_type == GuidedDecodingOptions.json_object:
+            ctx = self.compiler.compile_builtin_json_grammar()
+        elif request_type == GuidedDecodingOptions.grammar:
+            ctx = self.compiler.compile_grammar(grammar_spec)
+        else:
+            logger.error("Validation should have already occurred. "
+                         "Please file an issue.")
+            raise ValueError(
+                f"grammar is not of valid supported types. ({request_type!s})")
+
+        max_rollback_tokens = MAX_ROLLBACK_TOKENS
+        if self.vllm_config.speculative_config:
+            max_rollback_tokens = self.vllm_config.speculative_config\
+                                      .num_lookahead_slots + 1
+        return Grammar(matcher=xgr.GrammarMatcher(
+            ctx, max_rollback_tokens=max_rollback_tokens),
+                       vocab_size=self.vocab_size,
+                       ctx=ctx,
+                       max_rollback_tokens=max_rollback_tokens)
+
+    def setup_grammars(self):
```

Comment:
Dumb question: Why do we need this method?
Also, isn't it too expensive to call this every step?

Reply:
It's roughly "check all current structured output requests and set the grammar for any where the compilation is now complete."

I'll look into simplification / optimization.

Reply:
It turns out this was completely unnecessary, so I have removed it. I assume it was needed in some previous iteration.